pkg/auth/mtls_authhandler.go

```diff
@@ -0,0 +1,282 @@
+	tlsConn := tls.Client(conn, &tls.Config{
+		ServerName: m.cert.NumericIdentityToSNI(ar.remoteIdentity),
+		GetClientCertificate: func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return clientCert, nil
+		},
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: true, // not insecure as we do the verification in VerifyPeerCertificate
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			// verifiedChains will be nil as we set InsecureSkipVerify to true
+
+			chain := make([]*x509.Certificate, len(rawCerts))
+			for i, rawCert := range rawCerts {
+				cert, err := x509.ParseCertificate(rawCert)
+				if err != nil {
+					return fmt.Errorf("failed to parse certificate: %w", err)
+				}
+				chain[i] = cert
+			}
+
+			peerExpirationTime, err := m.verifyPeerCertificate(&ar.remoteIdentity, caBundle, [][]*x509.Certificate{chain})
+			if peerExpirationTime != nil && peerExpirationTime.Before(*expirationTime) {
+				expirationTime = peerExpirationTime // send down the lowest expiration time of the two certificates
+			}
+			return err
+		},
+		ClientCAs: caBundle,
+		RootCAs:   caBundle,
+	})
+	defer tlsConn.Close()
+
+	if err := tlsConn.Handshake(); err != nil {
+		return nil, fmt.Errorf("failed to perform TLS handshake: %w", err)
+	}
+
+	if expirationTime == nil {
+		return nil, fmt.Errorf("failed to get expiration time of peer certificate")
+	}
+
+	return &authResponse{
+		expirationTime: *expirationTime,
+	}, nil
+}
+
+func (m *mtlsAuthHandler) authType() policy.AuthType {
+	return policy.AuthTypeMTLSSpiffe
+}
+
+func (m *mtlsAuthHandler) listenForConnections(ctx context.Context, ready chan<- struct{}) {
+	// set up TCP listener
+
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", m.cfg.MTLSListenerPort))
+	if err != nil {
+		m.log.WithError(err).Fatal("Failed to start mTLS listener")
+	}
+	defer l.Close()
+
+	m.log.WithField(logfields.Port, m.cfg.MTLSListenerPort).Info("Started mTLS listener")
+	ready <- struct{}{} // signal to hive that we are ready to accept connections
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			conn, err := l.Accept()
+			if err != nil {
+				m.log.WithError(err).Error("Failed to accept connection")
+				if errors.Is(err, net.ErrClosed) {
+					m.log.Info("mTLS listener socket got closed")
+					return
+				}
+				continue
+			}
+			go m.handleConnection(ctx, conn)
+		}
+	}
+}
+
+func (m *mtlsAuthHandler) handleConnection(ctx context.Context, conn net.Conn) {
+	defer conn.Close()
+
+	caBundle, err := m.cert.GetTrustBundle()
+	if err != nil {
+		m.log.WithError(err).Error("failed to get CA bundle")
+		return
+	}
+
+	tlsConn := tls.Server(conn, &tls.Config{
+		ClientAuth:     tls.RequireAndVerifyClientCert,
+		GetCertificate: m.GetCertificateForIncomingConnection,
+		MinVersion:     tls.VersionTLS13,
+		ClientCAs:      caBundle,
+	})
+	defer tlsConn.Close()
+
+	if err := tlsConn.Handshake(); err != nil {
```

Comment:
We should honor the given context here
```suggestion
	if err := tlsConn.HandshakeContext(ctx); err != nil {
```